lib/ruby_lsp/server.rb

```diff
@@ -1,250 +1,158 @@
-      end
-    end
-
-    on("textDocument/documentLink", parallel: true) do |request|
-      uri = request.dig(:params, :textDocument, :uri)
-      store.cache_fetch(uri, :document_link) do |document|
-        RubyLsp::Requests::DocumentLink.new(uri, document).run
-      end
-    end
+    sig { returns(Thread) }
+    def new_worker
+      Thread.new do
+        # Thread::Queue#pop is thread safe and will wait until an item is available
+        loop do
+          job = T.let(@job_queue.pop, T.nilable(Job))
+          # The only time when the job is nil is when the queue is closed and we can then terminate the thread
+          break if job.nil?
 
-    on("textDocument/hover") do |request|
-      position = request.dig(:params, :position)
-      document = store.get(request.dig(:params, :textDocument, :uri))
+          request = job.request
+          @mutex.synchronize { @jobs.delete(request[:id]) }
 
-      RubyLsp::Requests::Hover.new(document, position).run
-    end
+          result = if job.cancelled
+            # We need to return nil to the client even if the request was cancelled
+            Result.new(response: nil, notifications: [])
+          else
+            Executor.new(@store).execute(request)
+          end
 
-    on("textDocument/foldingRange", parallel: true) do |request|
-      store.cache_fetch(request.dig(:params, :textDocument, :uri), :folding_ranges) do |document|
-        Requests::FoldingRanges.new(document).run
+          finalize_request(result, request)
+        end
       end
     end
 
-    on("textDocument/selectionRange", parallel: true) do |request|
-      uri = request.dig(:params, :textDocument, :uri)
-      positions = request.dig(:params, :positions)
-
-      ranges = store.cache_fetch(uri, :selection_ranges) do |document|
-        Requests::SelectionRanges.new(document).run
-      end
-
-      # Per the selection range request spec (https://microsoft.github.io/language-server-protocol/specification#textDocument_selectionRange),
-      # every position in the positions array should have an element at the same index in the response
-      # array. For positions without a valid selection range, the corresponding element in the response
-      # array will be nil.
-
-      unless ranges.nil?
-        positions.map do |position|
-          ranges.find do |range|
-            range.cover?(position)
-          end
+    # Finalize a Queue::Result. All IO operations should happen here to avoid any issues with cancelling requests
+    sig { params(result: Result, request: T::Hash[Symbol, T.untyped]).void }
+    def finalize_request(result, request)
+      @mutex.synchronize do
+        error = result.error
+        response = result.response
+
+        # If the response include any notifications, go through them and publish each one
+        notifications = result.notifications
+        notifications.each { |n| @writer.write(method: n.message, params: n.params) } if notifications.any?
```

Comment:
I feel the `any?` check is not necessary? If the `notifications` is empty, simply nothing would happen.


```suggestion
        result.notifications.each { |n| @writer.write(method: n.message, params: n.params) }
```